flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolServiceTest.java

```diff
@@ -0,0 +1,334 @@
+                    Collections.singletonList(
+                            new SlotOffer(new AllocationID(), 0, ResourceProfile.ANY));
+
+            final LocalTaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+
+            final Collection<SlotOffer> acceptedSlots =
+                    declarativeSlotPoolService.offerSlots(
+                            taskManagerLocation,
+                            new RpcTaskManagerGateway(
+                                    new TestingTaskExecutorGatewayBuilder()
+                                            .createTestingTaskExecutorGateway(),
+                                    jobMasterId),
+                            slotOffers);
+
+            assertThat(acceptedSlots, is(empty()));
+        }
+    }
+
+    @Test
+    public void testSlotOfferingOfKnownTaskManager() throws Exception {
+        final AtomicReference<Collection<? extends SlotOffer>> receivedSlotOffers =
+                new AtomicReference<>();
+        try (DeclarativeSlotPoolService declarativeSlotPoolService =
+                createDeclarativeSlotPoolService(
+                        new TestingDeclarativeSlotPoolFactory(
+                                new TestingDeclarativeSlotPoolBuilder()
+                                        .setOfferSlotsFunction(
+                                                (slotOffers,
+                                                        taskManagerLocation,
+                                                        taskManagerGateway,
+                                                        aLong) -> {
+                                                    receivedSlotOffers.set(slotOffers);
+                                                    return new ArrayList<>(slotOffers);
+                                                })))) {
+            final LocalTaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+
+            declarativeSlotPoolService.registerTaskManager(taskManagerLocation.getResourceID());
+
+            final Collection<SlotOffer> slotOffers =
+                    Collections.singletonList(
+                            new SlotOffer(new AllocationID(), 0, ResourceProfile.ANY));
+
+            declarativeSlotPoolService.offerSlots(
+                    taskManagerLocation,
+                    new RpcTaskManagerGateway(
+                            new TestingTaskExecutorGatewayBuilder()
+                                    .createTestingTaskExecutorGateway(),
+                            jobMasterId),
+                    slotOffers);
+
+            assertThat(receivedSlotOffers.get(), is(slotOffers));
+        }
+    }
+
+    @Test
+    public void testConnectToResourceManagerDeclaresRequiredResources() throws Exception {
+        final Collection<ResourceRequirement> requiredResources =
+                Arrays.asList(
+                        ResourceRequirement.create(ResourceProfile.ANY, 2),
```

Comment:
use `UNKNOWN` instead. `ANY` should only be used for the resource profile of a slot.
This isn't really about the correctness of this test in particular but more about not spreading code that would in other places lead to subtle issues.

Reply:
Very good point. Thanks a lot. I always get confused by `UNKNOWN` and `ANY`. I think it would be best to have two different types: 1) for the resource requirements and 2) for what the slot offers. Then we would also get rid of the problem with `ResourceProfile.isMatching` which confuses me every time in which order to supply the two `ResourceProfiles`.